### PR TITLE
Adds an endpoint that works as a facade for api-service-datawarehousereader and the required tests.

### PR DESCRIPTION
**Pull request by AndreSoderlund**

### CODE REVIEW

File: src/main/java/se/sundsvall/installation/api/InstallationResource.java [hunk 1]
```diff
@@ -20,5 +38,18 @@ class InstallationResource {
...
+	@Operation(summary = "Get uninstalled installations",
+		description = "Resource returns all uninstalled installations matching provided search parameters")
+	@ApiResponses(value = {
+		@ApiResponse(responseCode = "200", description = "Successful operation", useReturnTypeSchema = true),
+		@ApiResponse(responseCode = "400", description = "Bad request", content = @Content(mediaType = APPLICATION_PROBLEM_JSON_VALUE, schema = @Schema(oneOf = {Problem.class, ConstraintViolationProblem.class}))),
+		@ApiResponse(responseCode = "500", description = "Internal Server error", content = @Content(mediaType = APPLICATION_PROBLEM_JSON_VALUE, schema = @Schema(implementation = Problem.class))),
+		@ApiResponse(responseCode = "502", description = "Bad Gateway", content = @Content(mediaType = APPLICATION_PROBLEM_JSON_VALUE, schema = @Schema(implementation = Problem.class)))
+	})
+	@GetMapping(produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE})
+	public ResponseEntity<InstallationDetailsResponse> getInstallations(
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/api/InstallationResource.java:50]: Use own class for response from service, do not let DWR "bleed trough" (InstallationDetailsResponse).

File: src/main/java/se/sundsvall/installation/service/InstallationService.java [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
 
-	public InstallationService() {
-		// placeholder
+	private final DataWarehouseReaderClient dataWarehouseReaderClient;
+
+	public InstallationService(final DataWarehouseReaderClient dataWarehouseReaderClient) {
+		this.dataWarehouseReaderClient = dataWarehouseReaderClient;
+	}
+
+	public InstallationDetailsResponse getInstallations(final SearchParameters searchParameters) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/service/InstallationService.java:21]: See comment in API class

File: src/main/java/se/sundsvall/installation/service/mapper/RequestMapper.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+	private RequestMapper() {
+		//prevent instantiation
+	}
+
+	public static InstallationParameters toInstallationParameters(final SearchParameters searchParameters) {
+		return Optional.ofNullable(searchParameters).map(searchParams -> new InstallationParameters()
+				.installed(searchParams.getInstalled())
+				.dateFrom(searchParams.getDateFrom())
+				.category(Category.fromValue(searchParams.getCategory()))
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/service/mapper/RequestMapper.java:20]: IllegalArgumentException will be thrown if searchParams.getCategory() is null

File: pom.xml [hunk 4]
```diff
@@ -1,118 +1,123 @@
...
-				<configuration>
-					<excludes>
-						<exclude>**/se/sundsvall/installation/Application*</exclude>
-						<exclude>**/generated/se/sundsvall/**</exclude>
-					</excludes>
-				</configuration>
-			</plugin>
-		</plugins>
-	</build>
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply pom.xml:3]: I got a code style scheme from Jonas a couple of months ago that I'm using. Is this really an issue or are we just looking for things to comment on?

File: src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{
```
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json:1]: According to the story, the response should be identical to that of installedbase (minus the customer number). InstalledBase is not returning any paging information, hence this is neither.

File: src/test/java/se/sundsvall/installation/api/InstallationResourceTest.java [hunk 6]
```diff
@@ -1,4 +1,45 @@
...
+	void getInstallations() {
+		final var searchParameters = createSearchParameters();
+		final var parameterObject = createParameterMap(searchParameters);
+
+		webTestClient.get().uri(uriBuilder -> uriBuilder.path("/installations")
+				.queryParams(parameterObject).build())
+			.exchange()
+			.expectStatus().isOk();
+
+		verify(installationServiceMock).getInstallations(any(SearchParameters.class));
```
Comment (josc04) [inline src/test/java/se/sundsvall/installation/api/InstallationResourceTest.java:41]: Why not use a captor here to verify that the sent in parameter values are used when calling service?
```
	@Captor
	private ArgumentCaptor<SearchParameters> searchParametersCaptor;

	void getInstallations() {
		...
		verify(installationServiceMock).getInstallations(SearchParametersCaptor.capture());

  		assertThat(searchParametersCaptor.getValue()).satisfies(bean -> {
			assertThat(bean.getCategory()).isEqualTo(searchParameters.getCategory());
			assertThat(bean.getDateFrom()).isEqualTo(searchParameters.getDateFrom());
			assertThat(bean.getFacilityId()).isEqualTo(searchParameters.getFacilityId());
			assertThat(bean.getInstalled()).isEqualTo(searchParameters.getInstalled());
			assertThat(bean.getLimit()).isEqualTo(searchParameters.getLimit());
			assertThat(bean.getPage()).isEqualTo(searchParameters.getPage());
			assertThat(bean.getSortBy()).isEqualTo(searchParameters.getSortBy());
			assertThat(bean.getSortDirection()).isEqualTo(searchParameters.getSortDirection());
		});
```
(You'll discover that your createParameterMap doesn't handle lists corrrectly)

File: src/test/java/se/sundsvall/installation/service/InstallationServiceTest.java [hunk 7]
```diff
@@ -1,4 +1,35 @@
...
+	@InjectMocks
+	private InstallationService installationService;
+
+	@Test
+	void getInstallations() {
+		final var searchParameters = createSearchParameters();
+
+		installationService.getInstallations(searchParameters);
+
+		verify(dataWarehouseReaderClient).getInstallationDetails(any(InstallationParameters.class));
```
Comment (josc04) [inline src/test/java/se/sundsvall/installation/service/InstallationServiceTest.java:32]: Use captor here to verify that correct values are passed to the dataWarehouseReaderClient

File: src/integration-test/resources/InstallationIT/__files/test1_getInstallations/responses/dwr-response.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
+{
```
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply src/integration-test/resources/InstallationIT/__files/test1_getInstallations/responses/dwr-response.json:1]: Good catch, this is correct. DWR is returning paging information.

File: src/test/java/se/sundsvall/installation/service/mapper/RequestMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	@Test
+	void toInstallationParametersInvalidCategory() {
+		final var value = "INVALID";
+		final var searchParameters = createSearchParameters(p -> p.setCategory(value));
+
+		assertThatThrownBy(() -> RequestMapper.toInstallationParameters(searchParameters))
+			.isInstanceOf(IllegalArgumentException.class)
+			.hasMessage("Unexpected value '" + value + "'");
+	}
+
```
Comment (josc04) [inline src/test/java/se/sundsvall/installation/service/mapper/RequestMapperTest.java:37]: Suggestion is to add tests for null and empty parameter object:
```
	@Test
	void toInstallationParametersFromNull() {
		assertThat(RequestMapper.toInstallationParameters(null)).isNull();
	}

	@Test
	void toInstallationParametersFromEmptyParameters() {
		final var bean = RequestMapper.toInstallationParameters(SearchParameters.builder().build());

		assertThat(bean).hasAllNullFieldsOrProperties();
	}
```

File: pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (josc04) [reply pom.xml:3]: We have agreed on using tabs as indentation in the team. Not using it will make files indentation difficult when developers with different styles are editing the files. Hence it is a good thing to use the indentation style agreed upon.

File: src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (josc04) [reply src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json:1]: The story states:

> Att göra
> Ny fasadtjänst Installations som integrerar med ny metod i DataWarehouseReader
> Ny metod i DataWarehouseReader GET /installations
> Utred punkt 1 med Mikael F från Datalagret
> Lägg till paginering (kom ihåg att vi har olika sätt att paginera beroende på hur vyn är uppsatt)

File: src/main/java/se/sundsvall/installation/api/model/SearchParameters.java
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/installation/api/model/SearchParameters.java:1]: Not quite sure, in this case they are not used for pagination in this service they are just passed to DWR. So in my mind these are just like any other parameter.

File: src/main/java/se/sundsvall/installation/api/validation/impl/ValidCategoryConstraintValidator.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.List;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+import se.sundsvall.installation.api.validation.ValidCategory;
+
+public class ValidCategoryConstraintValidator implements ConstraintValidator<ValidCategory, String> {
+
+	public static final List<String> VALID_CATEGORIES = List.of("ELECTRICITY", "WASTE_MANAGEMENT", "ELECTRICITY_TRADE", "DISTRICT_COOLING", "DISTRICT_HEATING");
```
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/installation/api/validation/impl/ValidCategoryConstraintValidator.java:12]: The enum in DWR contains two additional values that are not listed as valid values in the story. Might be an oversight though.

File: src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndreSoderlund) [reply src/integration-test/resources/InstallationIT/__files/test1_getInstallations/expected.json:1]: This is just a facade that passes an incoming request to DWR, DWR in turn uses pagination based on the request parameters. When the "page" is returned the content is extracted and returned.

File: src/integration-test/java/apptest/InstallationIT.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+	@Test
+	void test1_getInstallations() throws JsonProcessingException, ClassNotFoundException {
+		setupCall()
+			.withServicePath("/installations?installed=false&facilityId=123&dateFrom=2022-01-01&category=ELECTRICITY&page=1&limit=10&sortBy=facilityId&sortDirection=ASC")
+			.withHttpMethod(GET)
+			.withExpectedResponse(EXPECTED_FILE)
+			.withExpectedResponseStatus(OK)
+			.sendRequestAndVerifyResponse()
+			.andReturnBody(InstallationDetailsResponse.class);
```
Comment (josc04) [inline src/integration-test/java/apptest/InstallationIT.java:30]: This is not needed, remove row 30.

File: src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package se.sundsvall.installation.api.model;
+
+import static io.swagger.v3.oas.annotations.media.Schema.AccessMode.READ_ONLY;
+
+import java.util.List;
+
+import generated.se.sundsvall.datawarehousereader.InstallationDetails;
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java:7]: This is still using DWR classes which it shouldn't. Create own classes to decouple this service from the underlying DWR service.

File: src/main/java/se/sundsvall/installation/api/model/SearchParameters.java
[the remark this answers is not in the document]
Comment (josc04) [reply src/main/java/se/sundsvall/installation/api/model/SearchParameters.java:1]: I think the pagination should be returned all the way to the clients of this service. How will the client know how many pages they receive if the information from DWR is not returned all the way. 

For example, if I as a client asks query that returns 1000 rows, and I limit the result to 100 in my request along with setting the page to 1, then I should get the information back that the result is totally 10 pages (with 100 rows in each page) and that I right now is recieving page 1. That way I as a client knows that there is more pages to read for me, if I want to receieve the full response to my query, right?

File: src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
+package se.sundsvall.installation.api.model;
+
+import static io.swagger.v3.oas.annotations.media.Schema.AccessMode.READ_ONLY;
+
+import java.util.List;
+
+import generated.se.sundsvall.datawarehousereader.PagingAndSortingMetaData;
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java:7]: DWR-class that should be replaced to decouple the facade class from the backend service. It's better to use the PagingMetaData class in Dept-44. See how it's done in api-service-document for example.
```
import se.sundsvall.dept44.models.api.paging.PagingMetaData;

@Schema(description = "Paged document response model", accessMode = READ_ONLY)
public class PagedDocumentResponse {

	@ArraySchema(schema = @Schema(implementation = Document.class))
	private List<Document> documents;

	@JsonProperty("_meta")
	@Schema(implementation = PagingMetaData.class)
	private PagingMetaData metadata;

(getters and setters omitted for brevity)
```

File: src/main/java/se/sundsvall/installation/api/model/MetaDataEmbeddable.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.ToString;
+
+
+@Getter
+@Setter
+@Builder(setterPrefix = "with")
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+public class MetaDataEmbeddable {
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/api/model/MetaDataEmbeddable.java:18]: To me the word `Embeddable` is connected to the DB model, maybe just call the class for InstallationMetaData?

File: src/test/java/se/sundsvall/installation/api/validation/impl/ValidPaginationConstraintValidatorTest.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+	private ConstraintValidatorContext context;
+
+	@Mock
+	private ConstraintValidatorContext.ConstraintViolationBuilder constraintViolationBuilder;
+
+	@Mock
+	private ConstraintValidatorContext.ConstraintViolationBuilder.NodeBuilderCustomizableContext nodeBuilderCustomizableContext;
+
+	@ParameterizedTest
+	@ValueSource(ints = {1, 50, 100})
```
Comment (josc04) [inline src/test/java/se/sundsvall/installation/api/validation/impl/ValidPaginationConstraintValidatorTest.java:32]: Change 100 to 1000 to test max valid limit

File: src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java (hunk 13 above)
Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java:7]: Used the wrong import..

File: src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+@Builder(setterPrefix = "with")
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+public class InstallationsResponse {
+
+	@ArraySchema(schema = @Schema(implementation = Installation.class, accessMode = READ_ONLY))
+	private List<Installation> installationDetails;
+
+	@Schema(description = "Paging and sorting metadata", accessMode = READ_ONLY)
```
Comment (josc04) [inline src/main/java/se/sundsvall/installation/api/model/InstallationsResponse.java:29]: Add `@JsonProperty("_meta")` to conform to DIGGs guidelines on response with paging and sorting (https://dev.dataportal.se/rest-api-profil/filtrering-paginering-och-sokparametrar)?

**Review by josc04 — APPROVED** [review]
